### PR TITLE
Encode & Decode Encrypted Records

### PR DESCRIPTION
**Pull request by m4dc4p**
Implemented methods for encoding encrypted records as strings and for decoding them from strings. 

* Added `EncryptedRecord` class (which implements `Record`) that has
methods for encoding and decoding.
* Updated `Client#encryptRecord` and `Client#encryptExisting` to return
`EncryptedRecord` instances, rather than the base `Record` interface.
* Updated `decryptExisting` to return a `LocalRecord` instance for concreteness.
* Fixed a bug where local encryption & decryption did noit require that the client have a signing
key. Necessitates a 3.0.0 release, as this is a backwards-incompatible change.

### CODE REVIEW

File: e3db/src/main/java/com/tozny/e3db/Client.java [hunk 1]
```diff
@@ -1588,82 +1588,69 @@ public void run() {
...
-    }
+  public LocalRecord decryptExisting(Record record, EAKInfo eakInfo) throws E3DBException {
+    if(record.signature() == null)
+      throw new E3DBVerificationException(record.meta(), "Record must have a signature in order to decrypt.");
 
-    if (record.signature() == null) {
-      throw new E3DBVerificationException(record.meta());
-    }
+    if (eakInfo.getSignerSigningKey() == null || eakInfo.getSignerSigningKey().isEmpty())
+      throw new IllegalStateException("eakInfo cannot be used to verify the record as it has no public signing key");
```
Comment (gstro) [inline e3db/src/main/java/com/tozny/e3db/Client.java]: Should this be a `E3DBVerificationException`?

File: e3db/src/main/java/com/tozny/e3db/Client.java [hunk 2]
```diff
@@ -1588,82 +1588,69 @@ public void run() {
...
     checkNotNull(record, "record");
     checkNotNull(eakInfo, "eak");
 
-    String signature = record.signature();
-    if (signature == null && Client.this.privateSigningKey != null) {
-      SignedDocument<Record> document = sign(record);
-      signature = document.signature();
-    }
+    if(Client.this.privateSigningKey == null)
+      throw new IllegalStateException("Client must have a signing key to encrypt locally.");
```
Comment (gstro) [inline e3db/src/main/java/com/tozny/e3db/Client.java]: Makes me think that we should somehow allow a "late" registration of public signing key, but IIRC the mobile SDKs landed with signing keys? Maybe not...

File: e3db/src/main/java/com/tozny/e3db/EncryptedRecord.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+public class EncryptedRecord implements Record {
+  private final Map<String, String> data;
+  private final RecordMeta meta;
+  private final String signature;
+
+  private static final ObjectMapper mapper;
+
+  static {
+    mapper = new ObjectMapper();
+    mapper.configure(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS, true);
```
Comment (gstro) [inline e3db/src/main/java/com/tozny/e3db/EncryptedRecord.java]: Handy.

File: e3db/src/main/java/com/tozny/e3db/EncryptedRecord.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+      if (record.meta().plain() != null)
+        meta.put("plain", mapper.writeValueAsString(record.meta().plain()));
+
+      meta.put("userId", record.meta().userId().toString());
+      meta.put("writerId", record.meta().writerId().toString());
+
+      HashMap<String, Object> r = new HashMap<>();
+      r.put("data", record.data());
+      r.put("meta", meta);
+      r.put("rec_sig", record.signature());
```
Comment (gstro) [inline e3db/src/main/java/com/tozny/e3db/EncryptedRecord.java]: Looks like you're mixing camel and snake cases?

File: e3db/src/main/java/com/tozny/e3db/EncryptedRecord.java [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+import com.fasterxml.jackson.databind.SerializationFeature;
+
+import java.io.IOException;
+import java.text.DateFormat;
+import java.util.*;
+
+import static com.tozny.e3db.Checks.checkNotNull;
+
+/**
+ * A record holding encrypted data, plus a signature. Should be used to persiste records to storage (and to
```
Comment (gstro) [inline e3db/src/main/java/com/tozny/e3db/EncryptedRecord.java]: Typo, should be _persist_

**Review by gstro — APPROVED** [review]
Looks good, just some minor details. Nice work.

File: androidtest/src/main/java/com/tozny/e3dbtest/KPLockScreenFragment.java [hunk 6]
```diff
@@ -55,7 +55,7 @@ public String configName() {
 
     @Override
     public KeyAuthentication keyProtection() {
-        return KeyAuthentication.withLockScreen();
+        return KeyAuthentication.withLockScreen(5);
```
Comment (ericmann) [inline androidtest/src/main/java/com/tozny/e3dbtest/KPLockScreenFragment.java:58]: Should we document what the value of `5` is doing?

File: e3db/src/main/java/com/tozny/e3db/EncryptedRecord.java (hunk 4 above)
Comment (ericmann) [reply e3db/src/main/java/com/tozny/e3db/EncryptedRecord.java]: We use snake case in all of our serialized stuff ... but often camel case in actual code. So long as we're intentional about how and where we're using different cases I think it's fine.

**Review by ericmann — APPROVED** [review]

File: androidtest/src/main/java/com/tozny/e3dbtest/KPLockScreenFragment.java (hunk 6 above)
Comment (m4dc4p) [reply androidtest/src/main/java/com/tozny/e3dbtest/KPLockScreenFragment.java:58]: I don't think it's necessary - the IDE should show docs, or they can jump to source easily. In any case this is just the test app.

File: e3db/src/main/java/com/tozny/e3db/EncryptedRecord.java (hunk 4 above)
Comment (m4dc4p) [reply e3db/src/main/java/com/tozny/e3db/EncryptedRecord.java]: I updated in a later commit in any case.

File: androidtest/src/main/java/com/tozny/e3dbtest/KPLockScreenFragment.java (hunk 6 above)
Comment (ericmann) [reply androidtest/src/main/java/com/tozny/e3dbtest/KPLockScreenFragment.java:58]: :+1:

**Review by ericmann — APPROVED** [review]

**Comment by m4dc4p** [comment]
Re-submitted as #18.

**Pull request closed without merging**